### ISSUE TITLE
ESLint & Developer Tools

### ISSUE THREAD

**Issue by Standard8** [devtools-html/rfcs#30]
Currently devtools/ and the rest of mozilla-central have two fairly different configurations for ESLint.

Whilst the majority of mozilla-central is covered by [eslint-plugin-mozilla's configurations](https://searchfox.org/mozilla-central/search?q=&case=false&regexp=false&path=eslint-plugin-mozilla%2Flib%2Fconfigs%2F), devtools defines most of its [own rules](https://searchfox.org/mozilla-central/search?q=&case=false&regexp=false&path=devtools%2F**.eslintrc*.js), even if they are a duplicate of the main ones.

I would like to simplify the configurations so that it is clearer what devtools does differently, and that devtools can more easily benefit from changes/improvements to the mozilla configurations. One of the eventual aims would be to unify the configurations across mozilla-central to make it easier for developers to move from one to the other.

I'm proposing the following:

* We aim to simplify the devtools ruleset so that it is primarily derived from mozilla/recommended & the related test configurations.
* We remove duplicate rules where they are the same as the mozilla configs.
* For now, we leave rules that are different or extra to the mozilla configurations (let's think about harmonising those later).
* The devtools.html repositories could extend the eslint-plugin-mozilla configuration (note: this gets [published via npm](https://www.npmjs.com/package/eslint-plugin-mozilla)), with the aim of simplifying rules in those as well.

I'm happy to lead the effort with filing bugs, and doing a little work (e.g. de-dupe configs), though I'll likely also use mentored bugs.

**Comment by Standard8** [comment]
Just a bit of history. Sometime earlier this year I had discussed this with someone (sorry I can't remember who), and it was suggested that with the possible move to system add-ons/separate repos, that we leave the devtools configuration alone at the time.

I believe no move is currently in the plan, but even if it is, I think this is still worth doing and deriving from the mozilla config - that can still be used outside of mozilla-central, and it provides more consistency for developers.

**Comment by juliandescottes** [comment]
Thanks for opening this as a RFC. Proposal looks good to me.

> I believe no move is currently in the plan, but even if it is, I think this is still worth doing and deriving from the mozilla config

Agreed. Plans for DevTools as an addon are on hold, so no need to wait on this before discussing this eslint cleanup.

**Comment by julienw** [comment]
Just a note about what we use in the devtools-html world. We often use [prettier](https://github.com/prettier/prettier) and so we don't enable stylistic rules from eslint. I believe this would be the biggest issue. I consider now that configuring prettier was a very big win in how our linting was done.

See for example the eslintrc file we use in perf.html: https://github.com/julienw/perf.html/blob/23af7edc23ae13d90b5ff7bb799daad949dec3c1/.eslintrc.js (this specific one, with some added rules for react, should land today).

I also think that we're more opinionated, and we can be especially because when enabling stricter rules we don't need to fix a lot of existing code. Although that could be done even when extending mozilla's recommended rules.

**Comment by Standard8** [comment]
> Just a note about what we use in the devtools-html world. We often use prettier and so we don't enable stylistic rules from eslint. I believe this would be the biggest issue. I consider now that configuring prettier was a very big win in how our linting was done.

Those can be some of the exceptions to the Mozilla rules (I will try and think of a simpler file). I don't have any sense of other people's ideas yet, but one idea that I've had is once clang-format is in use for m-c for the c++ code, then maybe we should consider something more for js as well, e.g. prettier.

Just an additional clarification whilst I'm here: We don't necessarily need to apply the mozilla config to all of devtools' repositories. We probably only want to do it for the ones that are incorporated into gecko code. This is because the recommended config assumes gecko.

I guess it might be that we need to split that config into two - one for the basic rules for eslint that we generally want to have across Mozilla, the other for Mozilla specific gecko items.

**Comment by julienw** [comment]
Yes, I also noticed globals we don't have normally in web development :) It makes sense to have a set of basic generic JS rules. Feel free to reuse ours ;)

**Comment by gregtatum** [comment]
I do like the idea of being able to npm install rules for some base guidelines that I can easily bring into any project, and echoing Julien, I don't want it to conflict with our Prettier configuration. I'm not very opinionated as to the actual implementation.

**Comment by juliandescottes** [comment]
> Just an additional clarification whilst I'm here: We don't necessarily need to apply the mozilla config to all of devtools' repositories. We probably only want to do it for the ones that are incorporated into gecko code. This is because the recommended config assumes gecko.

+1, that was my assumption as well. Looking at DevTools repositories on GitHub, I think only one eslintrc file is shared between GitHub and mozilla-central: [eslintrc for debugger.html mochitests](https://github.com/devtools-html/debugger.html/blob/master/src/test/mochitest/.eslintrc)

For all the other repositories (devtools-core, perf-html, the rest of debugger.html), it's probably best to leave this up to the maintainers. If the repo maintains code that might be moved to or from mozilla-central (such as some devtools-core packages maybe?), remaining close to the mozilla-central rules can make sense.

**Comment by juliandescottes** [comment]
My comment was apparently lost! @Standard8 , this was accepted, sorry I didn't ping you earlier. Feel free to start logging bugs to update the devtools eslint rules.

**Comment by Standard8** [comment]
Thanks Julian, I've filed [bug 1430596](https://bugzilla.mozilla.org/show_bug.cgi?id=1430596) as the meta to track the work here.

**Comment by juliandescottes** [comment]
Great! Closing, the work will happen under the meta bug mentioned above.

**Comment by juliandescottes** [comment]
Last bug has been resolved, on top of that we have a few follow ups under the following meta: 
https://bugzilla.mozilla.org/show_bug.cgi?id=1467767
